File: 450-delete-node-in-a-bst/delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:

        if not root: return None

        if key < root.val:
            root.left = self.deleteNode(root.left,key)
        elif key > root.val:
            root.right = self.deleteNode(root.right,key)
        
        else :
            if not root.right: return root.left
            elif not root.left : return root.right
            else :
                #finding mining val
                min_Val = self.findMin(root.right)
                # replacing it 
                root.val = min_Val.val
                # delete child which is replaced as root 
                root.right = self.deleteNode(root.right,min_Val.val)
        return root 


    def findMin(self,root):
        curr = root
        while curr and curr.left:
            curr = curr.left
        return curr

File: 450-delete-node-in-a-bst/delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (!root){return nullptr;}
        if (root->val >key){root->left = deleteNode(root->left, key);}
        else if(root->val < key){root->right = deleteNode(root->right,key);}
        else{
        if(!root->left){return root->right;}
        if(!root->right){return root->left;}
 //if left right persent then find min of right replace it node and kill the child
         TreeNode* curr = root->right;
            while (curr->left) {
                curr = curr->left;
            }
            root->val = curr->val;
            root->right = deleteNode(root->right, root->val);
    }
    return root;
    }
};